File: Remove_Invalid_Parentheses.cpp
/*301. Remove Invalid Parentheses  QuestionEditorial Solution  My Submissions
Total Accepted: 18009
Total Submissions: 53928
Difficulty: Hard
Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results.

Note: The input string may contain letters other than the parentheses ( and ).

Examples:
"()())()" -> ["()()()", "(())()"]
"(a)())()" -> ["(a)()()", "(a())()"]
")(" -> [""]
Credits:
Special thanks to @hpplayer for adding this problem and creating all test cases.

Subscribe to see which companies asked this question*/



/*define left and right to decide if the parentheses are a pair*/
/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        unordered_set<string> ret;
        int left = 0;
        int right = 0;
        for(auto c:s){
            if(c == '('){
                ++left;
            }
            
            if(c == ')'){
                if(left != 0){
                    --left;
                } else {
                    ++right;
                }
            }
        }
        
        check(s, 0, left, right, 0, "", ret);
        return vector<string>(ret.begin(), ret.end());
    }
    
private:
    void check(string s, int index, int left, int right, int pair, string path, unordered_set<string>& ret){
        if(index == s.size()){
            if(left == 0 && right == 0 && pair == 0){
                ret.insert(path);
            }
            return;
        }
        
        if(s[index] != '(' && s[index] != ')'){
            check(s, index + 1, left, right, pair, path+s[index], ret);
        } else {
            if(s[index] == '('){
                if(left > 0){
                    check(s, index + 1, left - 1, right, pair, path, ret);
                }
                check(s, index + 1, left, right, pair + 1, path + s[index], ret);
            }
            
            if(s[index] == ')'){
                if(right > 0){
                    check(s, index + 1, left, right - 1, pair, path, ret);
                }
                
                if(pair > 0){
                    check(s, index + 1, left, right, pair - 1, path + s[index], ret);
                }
            }
        }
    }
};





/*Use one count to decide if the parentheses are pairs*/
/*Time complexity:

In BFS we handle the states level by level, in the worst case, we need to handle all the levels, we can analyze the time complexity level by level and add them up to get the final complexity.

On the first level, there's only one string which is the input string s, let's say the length of it is n, to check whether it's valid, we need O(n) time. On the second level, we remove one ( or ) from the first level, so there are C(n, n-1) new strings, each of them has n-1 characters, and for each string, we need to check whether it's valid or not, thus the total time complexity on this level is (n-1) x C(n, n-1). Come to the third level, total time complexity is (n-2) x C(n, n-2), so on and so forth...

Finally we have this formula:

T(n) = n x C(n, n) + (n-1) x C(n, n-1) + ... + 1 x C(n, 1) = n x 2^(n-1).*/

/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public List<String> removeInvalidParentheses(String s) {
        List<String> ret = new ArrayList<>();
        if(s==null){
            return ret;
        }
        
        Set<String> S = new HashSet<>();
        Queue<String> Q = new LinkedList<>();
        
        Q.add(s);
        S.add(s);
        
        boolean found = false;
        
        while(!Q.isEmpty()){
            s = Q.poll();
            if(isValid(s)){
                ret.add(s);
                found = true;
            }
            
            if(found) continue;
            
            for(int i = 0; i < s.length(); i++){
                if(s.charAt(i) != '(' && s.charAt(i) != ')') continue;
                String t = s.substring(0,i) + s.substring(i+1);
                if(!S.contains(t)){
                    Q.add(t);
                    S.add(t);
                }
            }
        }
        return ret;
    }
    
    public boolean isValid(String s){
        int count = 0;
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(c == '(') count++;
            // System.out.println(count);
            if(c == ')' && count-- == 0) return false;
        }
        return count == 0;
    }
}

File: Remove_Invalid_Parentheses.swift
/*
Remove the minimum number of invalid parentheses in order to make the input string valid. Return all possible results.

Note: The input string may contain letters other than the parentheses ( and ).

Example 1:

Input: "()())()"
Output: ["()()()", "(())()"]
Example 2:

Input: "(a)())()"
Output: ["(a)()()", "(a())()"]
Example 3:

Input: ")("
Output: [""]
*/

/*
Solution 2:
limited backTrack

find number of misplaced ( and )
- We process the expression one bracket at a time starting from the left.
- Suppose we encounter an opening bracket i.e. (, it may or may not lead to an invalid expression because there can be a matching ending bracket somewhere in the remaining part of the expression. Here, we simply increment the counter keeping track of left parentheses till now. left += 1
- If we encounter a closing bracket, this has two meanings:
  - Either there was no matching opening bracket for this closing bracket and in that case we have an invalid expression. This is the case when left == 0 i.e. when there are no unmatched left brackets available. In such a case we increment another counter say right += 1 to represent misplaced right parentheses.
  - Or, we had some unmatched opening bracket available to match this closing bracket. This is the case when left > 0. In this case we simply decrement the left counter we had i.e. left -= 1
- Continue processing the string until all parentheses have been processed.
- In the end the values of left and right would tell us the number of unmatched ( and ) parentheses respectively.

Algorithm
- The state of the recursion is now defined by five different variables:
  - index which represents the current character that we have to process in the original string.
  - left_count which represents the number of left parentheses that have been added to the expression we are building.
  - right_count which represents the number of right parentheses that have been added to the expression we are building.
  - left_rem is the number of left parentheses that remain to be removed.
  - right_rem represents the number of right parentheses that remain to be removed. Overall, for the final expression to be valid, left_rem == 0 and right_rem == 0.
- When we decide to not consider a parenthesis i.e. delete a parenthesis, be it a left or a right parentheses, we have to consider their corresponding remaining counts as well. This means that we can only discard a left parentheses if left_rem > 0 and similarly for the right one we will check for right_rem > 0.
- There are no changes to checks for considering a parenthesis. Only the conditions change for discarding a parenthesis.
- Condition for an expression being valid in the base case would now become left_rem == 0 and right_rem == 0. Note that we don't have to check if left_count == right_count anymore because in the case of a valid expression, we would have removed all the misplaced or invalid parenthesis by the time the recursion ends. So, the only check we need if left_rem == 0 and right_rem == 0.

Time Complexity: O(2^n)
Space Complexity: O(n)
*/
class Solution {
    func removeInvalidParentheses(_ s: String) -> [String] {
        if s.isEmpty { return [""] }
        
        var s = s
        
        var left = 0
        var right = 0
        
        for c in s {
            if c == "(" {
                left += 1
            } else if c == ")" {
                if left == 0 {
                    right += 1
                } else {
                    left -= 1
                }
            }
        }
        
        var cur = String()
        
        // use set to avoid duplicate
        var valid = Set<String>()
        backTrack(s, s.startIndex, 0, 0, left, right, &cur, &valid)        
        return Array(valid)
    }
    
    func backTrack(_ s: String, _ index: String.Index, 
                   _ leftCount: Int, _ rightCount: Int, 
                   _ misLeftRemain: Int, _ misRightRemain: Int,
                   _ cur: inout String, _ valid: inout Set<String>) {
        if index == s.endIndex {
            if misLeftRemain == 0, misRightRemain == 0 {
                valid.insert(cur)
            }
        } else {
            let c = s[index]
            let count = cur.count
            
            if (c == "(" && misLeftRemain > 0) 
            || (c == ")" && misRightRemain > 0) {
                backTrack(s, 
                          s.index(after: index), 
                          leftCount, 
                          rightCount, 
                          c == "(" ? misLeftRemain-1 : misLeftRemain,
                          c == ")" ? misRightRemain-1 : misRightRemain,
                          &cur, &valid)
            }
            cur.append(c)
            
            if c != "(", c != ")" {
                backTrack(s, 
                          s.index(after: index), 
                          leftCount, 
                          rightCount, 
                          misLeftRemain,
                          misRightRemain,
                          &cur, &valid)
            } else if c == "(" {
                backTrack(s, 
                          s.index(after: index), 
                          leftCount+1, 
                          rightCount, 
                          misLeftRemain,
                          misRightRemain,
                          &cur, &valid)
            } else if rightCount < leftCount {
                backTrack(s, 
                          s.index(after: index), 
                          leftCount, 
                          rightCount+1, 
                          misLeftRemain,
                          misRightRemain,
                          &cur, &valid)
            }
            
            cur.removeLast()
        }
    }
}

/*
Soluton 1:
bfs

use visited to check already checked string
use queue to memo next checking string list

use isFound to help terminate removing char from valid string

Time Complexity: O(2^n)
Spacee Complexity: O(n)
*/
class Solution {
    func removeInvalidParentheses(_ s: String) -> [String] {
        if s.isEmpty { return [""] }
        
        var s = s
        
        var visited: Set<String> = [s]
        var queue: [String] = [s]
        
        var isFound = false
        
        var res = [String]()
        while !queue.isEmpty {
            s = queue.removeFirst()
            
            if isValid(s) {
                res.append(s)
                isFound = true
            }
            
            if isFound {
                continue
            }
            
            for i in s.indices {
                if s[i] != "(", s[i] != ")" {
                    continue
                }
                
                var temp = s
                temp.remove(at: i)
                if !visited.contains(temp) {
                    visited.insert(temp)
                    queue.append(temp)
                }
            }
        }
        return res
    }
    
    func isValid(_ s: String) -> Bool {
        if s.isEmpty { return true }
        
        var count = 0
        for c in s {
            if c == "(" {
                count += 1
            } else if c == ")" {
                count -= 1
                if count < 0 {
                    return false 
                }
            }
        }
        return count == 0
    }
}
